### PR TITLE
[query] Change EmitCode implementation

### PR DESCRIPTION
**Pull request by patrick-schultz**
This replaces the EmitCode implementation by a start label together with an IEmitCode. I propose that EmitCode will still be needed occasionally in a CodeBuilder world, and that this is a form it could take.

For compatibility with existing EmitCode users, this defines `setup` and `m`, but those should be removed after the transition to CodeBuilder/IEmitCode is complete.

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala:396]: I suspect this may break tests. Do all users of EmitCode always check missingness?

**Review by tpoterba — CHANGES_REQUESTED** [review]
This is a great change! Waiting for tests to pass before approval.

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala
Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala:396]: Ah, you're probably right. Hmm...

**Comment by patrick-schultz** [comment]
Tests are passing, but I'd like to see if I can go back to not having setup on EmitCode, now that we found the btree problem.

**Comment by patrick-schultz** [comment]
Nice, it works now! I flipped through uses of `EmitCode.setup`, looking for places that might cause problems when `m` was more than a load of a boolean. I only found a couple, and fixed them. Also note that `EmitCode.fromI` also makes an empty `setup`, with all the work done in `m`, so we were already assuming `m` can do work.

**Comment by tpoterba** [comment]
excellent! can you just back out the change to AppendOnlyBTree?

File: hail/src/main/scala/is/hail/expr/ir/agg/AppendOnlyBTree.scala
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/agg/AppendOnlyBTree.scala:241]: remove

**Review by tpoterba — CHANGES_REQUESTED** [review]

**Review by tpoterba — APPROVED** [review]

**Pull request merged**
